File: apprefurbished/src/main/java/com/androidcollider/easyfin/transaction_categories/root/TransactionCategoriesRootModel.java
package com.androidcollider.easyfin.transaction_categories.root;

import android.support.v4.util.Pair;

import com.androidcollider.easyfin.common.models.TransactionCategory;
import com.androidcollider.easyfin.common.repository.Repository;

import java.util.List;

import io.reactivex.Flowable;

/**
 * @author Ihor Bilous
 */

class TransactionCategoriesRootModel implements TransactionCategoriesRootMVP.Model {

    private Repository repository;


    TransactionCategoriesRootModel(Repository repository) {
        this.repository = repository;
    }


    @Override
    public Flowable<Pair<List<TransactionCategory>, List<TransactionCategory>>> getAllTransactionCategories() {
        return Flowable.combineLatest(
                repository.getAllTransactionIncomeCategories(),
                repository.getAllTransactionExpenseCategories(),
                Pair::new
        );
    }

    @Override
    public Flowable<TransactionCategory> addNewTransactionCategory(TransactionCategory transactionCategory, boolean isExpense) {
        return isExpense ?
                        repository.addNewTransactionExpenseCategory(transactionCategory) :
                        repository.addNewTransactionIncomeCategory(transactionCategory);
    }
}

File: apprefurbished/src/main/java/com/androidcollider/easyfin/transaction_categories/root/TransactionCategoriesRootModel.kt
package com.androidcollider.easyfin.transaction_categories.root

import com.androidcollider.easyfin.common.models.TransactionCategory
import com.androidcollider.easyfin.common.repository.Repository
import io.reactivex.rxjava3.core.Single

/**
 * @author Ihor Bilous
 */
internal class TransactionCategoriesRootModel(private val repository: Repository) :
    TransactionCategoriesRootMVP.Model {

    override val allTransactionCategories:
            Single<Pair<List<TransactionCategory>, List<TransactionCategory>>>
        get() = Single.zip(
            repository.allTransactionIncomeCategories!!,
            repository.allTransactionExpenseCategories!!
        ) { first: List<TransactionCategory>, second: List<TransactionCategory> ->
            Pair(first, second)
        }

    override fun addNewTransactionCategory(
        transactionCategory: TransactionCategory,
        isExpense: Boolean
    ): Single<TransactionCategory> {
        return if (isExpense) repository.addNewTransactionExpenseCategory(transactionCategory)
        else repository.addNewTransactionIncomeCategory(transactionCategory)
    }
}